google/cloud/pubsub/internal/tracing_message_batch.cc

```diff
@@ -40,8 +40,40 @@ void TracingMessageBatch::SaveMessage(pubsub::Message m) {
   child_->SaveMessage(std::move(m));
 }
 
-// TODO(#12528): Implement functionality for Flush.
-void TracingMessageBatch::Flush() { child_->Flush(); }
+void TracingMessageBatch::Flush() {
+  using opentelemetry::trace::SpanContext;
+  using AttributesList =
+      std::vector<std::pair<opentelemetry::nostd::string_view,
+                            opentelemetry::common::AttributeValue>>;
+  int64_t message_count = message_spans_.size();
+  auto constexpr kMaxOtelLinks = 128;
+  std::vector<std::pair<SpanContext, AttributesList>> links;
+
+  // If the batch size is less than the max size, add the links to a single
+  // span.
+  if (message_count < kMaxOtelLinks) {
+    for (int64_t i = 0; i < message_count; i++) {
+      auto span = message_spans_.at(static_cast<int>(i));
+      AttributesList link_attributes = {{"messaging.pubsub.message.link", i}};
+      links.emplace_back(std::make_pair(span->GetContext(), link_attributes));
+    }
+  }
+  auto batch_sink_span_parent = internal::MakeSpan(
+      "BatchSink::AsyncPublish",
+      /*attributes=*/
+      {{"messaging.pubsub.num_messages_in_batch", message_count}},
+      /*links*/ links);
+
+  // Clear message spans.
+  message_spans_.clear();
```

Comment:
I know the intent is to handle the case with more than 128 links in a future PR, but maybe for now we should just add the *first* 128 links instead of adding zero links when there are more than 128?

Reply:
In the impl for +128 links, it will not add any links to this span. Instead this span will become the parent for the children span, so leaving it as is.

Added comments about future implementation and a test to show this behavior is WAI.